### PR TITLE
fix: Debug and warning messages

### LINKED ISSUE #336: spglib-2.1.0 shows warnings when using `get_spacegroup` method but spglib-2.0.2 does not
The spglib is an useful tool to analysis the symmetry information for giving structure.

Here we want to use the spglib to analysis the symmetry info of given structure by `spglib.get_spacegroup` and `spglib.refine_cell` method in some different tolerances including 0.00001, 0.0001, 0.001, 0.01, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, but in spglib-2.1.0, it shows this: 

```
spglib: No point group was found (line 405, /tmp/pip-install-wo87okks/spglib/src/pointgroup.c).
spglib: Attempt 0 tolerance = 1.000000e-03 failed(line 800, /tmp/pip-install-wo87okks/spglib/src/spacegroup.c).
spglib: No point group was found (line 405, /tmp/pip-install-wo87okks/spglib/src/pointgroup.c).
spglib: Attempt 1 tolerance = 9.500000e-04 failed(line 800, /tmp/pip-install-wo87okks/spglib/src/spacegroup.c).
spglib: No point group was found (line 405, /tmp/pip-install-wo87okks/spglib/src/pointgroup.c).
spglib: Attempt 2 tolerance = 9.025000e-04 failed(line 800, /tmp/pip-install-wo87okks/spglib/src/spacegroup.c).
spglib: No point group was found (line 405, /tmp/pip-install-wo87okks/spglib/src/pointgroup.c).
spglib: Attempt 3 tolerance = 8.573750e-04 failed(line 800, /tmp/pip-install-wo87okks/spglib/src/spacegroup.c).
```
but in older version such as spglib-2.0.2, this warning does not show

Here is the structure info and test script which can be used to repeat the warning above:

```
BN
1.0
  2.13478593     -0.00047962     -0.00000019
 -1.06780826      1.84853908     -0.00000019
 -0.00000004     -0.00000007      1.94254452
 B N
 1 1
direct
0.00002343  0.99997657  0.50000000
0.66667658  0.33332342  0.00000000
```
```
import spglib as spg
from ase.io import read

version = ".".join(map(str, spg.get_version()))
print(f"spg version: {version}")

atoms = read("POSCAR")
cell = atoms.cell
position = atoms.get_scaled_positions()
number = atoms.get_atomic_numbers()
print(spg.get_spacegroup((cell, position, number)))
print(spg.get_spacegroup((cell, position, number), 0.00001))
print(spg.get_spacegroup((cell, position, number), 0.0001))
print(spg.get_spacegroup((cell, position, number), 0.001))
print(spg.get_spacegroup((cell, position, number), 0.01))
print(spg.get_spacegroup((cell, position, number), 0.1))
print(spg.get_spacegroup((cell, position, number), 0.2))
print(spg.get_spacegroup((cell, position, number), 0.3))
print(spg.get_spacegroup((cell, position, number), 0.4))
print(spg.get_spacegroup((cell, position, number), 0.5))
print(spg.get_spacegroup((cell, position, number), 0.6))
print(spg.get_spacegroup((cell, position, number), 0.7))
print(spg.get_spacegroup((cell, position, number), 0.8))

```

Is this a bug? or something that i did to occur this? 

Looking forward to a reply, thx~

### LINKED ISSUE #338: Most of warning_print should should not be displayed.
As @LecrisUT mentions at #336, `warning_print` is on by default. But I think most of them should not be displayed by default, but ideally this should be conditional. @LecrisUT, do you know a good way to make it conditional?

### PR DESCRIPTION
**Pull request by LecrisUT**
- [x] Refactor message implementation. Avoid macros whenever possible
- [x] Enable/Disable messages by environment variable. `SPGLIB_DEBUG` and `SPGLIB_WARNINGS` cmake variables are for controlling if these messages are compiled at all or not
- [x] Write appropriately to `stdout` for debug messages and `stderr` for warning
- [x] Review what messages are warnings and which ones are debug messages
  - [x] Check that warning messages are not associated with `attempt`
  - [ ] Check that warning messages are associated with a failure/error 
  - [ ] Add `warning_print`/`info_print` at the end of `for (attempt)` loops

Closes #338, Closes #336

Depends-on: #431

### CODE REVIEW

File: src/pointgroup.c [hunk 1]
```diff
@@ -401,8 +401,7 @@ Pointgroup ptg_get_transformation_matrix(int transform_mat[3][3],
         get_axes(axes, pointgroup.laue, &pointsym, aperiodic_axis);
         set_transformation_matrix(transform_mat, axes);
     } else {
-        warning_print("spglib: No point group was found ");
-        warning_print("(line %d, %s).\n", __LINE__, __FILE__);
+        warning_print("spglib: No point group was found\n");
```
Comment (LecrisUT) [inline src/pointgroup.c]: This one is a good example of the things that need to be reviewed:

From https://github.com/spglib/spglib/issues/338#issuecomment-2023414189, it seems that this message is one of those that is not related to an error. Thus in a normal run, this message gets spammed a lot. Do you think it should be changed to a debug message?

**Comment by LecrisUT** [comment]
I took a more conservative approach in b7adcc0fdf1425f795b47f4f0ea2a3ea750d5e69. It is unfortunate that some messages appear both in a lopp and in the main `spglib.h` interface so it's not a clear cut how to categorize them. I still think I got these messages mislabeled. We could go even more conservative and only address the messages that were reported by users, but that would take some time and bug reports to get it right.

File: src/pointgroup.c (hunk 1 above)
Comment (kavanase) [reply src/pointgroup.c]: Fwiw, in my opinion this message should be changed to a debug message, that ideally is only printed if the overall routine fails or the code is being run in some debug type mode. Thanks for looking into this @LecrisUT!

File: src/debug.c [hunk 2]
```diff
@@ -32,39 +32,120 @@
...
+    // it is defined, including SPGLIB_DEBUG=""
+    return true;
+}
+bool warning_enabled() {
+    char const* warning_env = getenv("SPGLIB_WARNING");
+    // If SPGLIB_WARNING is not defined assume warning is on
+    if (warning_env == NULL) return true;
+    // Check if SPGLIB_WARNING is disabled. Not performing case-insensitive
+    // checks.
+    if (strcmp(warning_env, "OFF") == 0) return false;
```
Comment (lan496) [inline src/debug.c:57]: Why only `warning_enabled` accepts "OFF" compared to `debug_enabled` and `info_enabled`?

File: src/determination.c [hunk 3]
```diff
@@ -77,9 +77,8 @@ DataContainer *det_determine_all(Cell const *cell, int const hall_number,
                          container->primitive->tolerance)) != NULL) {
                 goto found;
             }
-            warning_print(
-                "spglib: ref_get_exact_structure_and_symmetry failed.");
-            warning_print(" (line %d, %s).\n", __LINE__, __FILE__);
+            debug_print(
```
Comment (lan496) [inline src/determination.c:80]: An algorithm failure should be treated as info.

File: src/niggli.c [hunk 4]
```diff
@@ -348,7 +343,7 @@ static int step2(NiggliParams *p) {
 static int step2_for_layer(NiggliParams *p) {
     if (p->B > p->C + p->eps || (!(fabs(p->B - p->C) > p->eps) &&
                                  fabs(p->eta) > fabs(p->zeta) + p->eps)) {
-        warning_print(
+        debug_print(
```
Comment (lan496) [inline src/niggli.c:346]: It looks like warning

File: src/debug.c (hunk 2 above)
Comment (LecrisUT) [reply src/debug.c:57]: The others are `default: false`, while this is the only one to be `default: true`. For default false, we can assume if you set an environment variable the intent is to enable it. For the opposite case it is more ambiguous, that's why the explicit check of the value.

**Review by lan496 — COMMENTED** [review]
I assume the fixes in https://github.com/spglib/spglib/pull/461 will be included in this   PR.

**Comment by LecrisUT** [comment]
> I assume the fixes in https://github.com/spglib/spglib/pull/461 will be included in this PR.

I have cherry-picked only select changes from there, because I am unsure about converting all `warning_print` to `info_print`. If you find some other that need porting, feel free to mention the file-line.

I have also added another item on the list "Add `warning_print`/`info_print` at the end of `for (attempt)` loops". Could use some help with those.

**Review by lan496 — APPROVED** [review]
I'm **not** so opinionated in info or warning and the refactoring looks good to me.

**Review by atztogo — APPROVED** [review]

**Comment by kavanase** [comment]
Thanks for your work on this!
Just to flag @LecrisUT in case it flies under the radar, with `spglib 2.4.0` and updating `DYLD_LIBRARY_PATH` (so `spglib.spg_get_version_full()` shows the correct new version), most of the verbose warning messages are no longer printed (🙌 ), but these two are still printed:
```
spglib: ssm_get_exact_positions failed.
spglib: get_bravais_exact_positions_and_lattice failed.
```

I tried and these can't be captured with `warnings` or `StringIO`. They disappear when the same code is run again.

**Comment by LecrisUT** [comment]
You can set `SPGLIB_WARNING=OFF` as an environment variable. It should work within Python as well as long as a clean shell environment is not created every time

**Comment by kavanase** [comment]
Ok great, thanks very much for this!

One last question if you don't mind, is there any way to ensure that when users upgrade `spglib`, that it uses the correct `DYLD_LIBRARY_PATH` (without having to manually set it as I did with `export DYLD_LIBRARY_PATH=/path/to/site-packages/spglib/lib:$DYLD_LIBRARY_PATH`)?

**Comment by LecrisUT** [comment]
That I am still thinking about in #462. In the meantime, you can try:
```console
$ pip install spglib --config-settings=cmake.define.SPGLIB_SHARED_LIBS=OFF
```

I am not sure if it works with `pip install spglib`, but it should work with `pip install git+https://github.com/spglib/spglib`. This one links the `_spglib` library statically to the bundled version

Other options would be to `patchelf --add-rpath=$ORIGIN/libs` (not sure if I got the path correct or if `$ORIGIN` needs to be escaped) to the `_spglib` library. `RPATH` should have priority when linking, and adding this would be like specifying `DYLD_LIBRARY_PATH` statically. `patchelf` should be available for MacOS in `homebrew` :shrug:

**Comment by kavanase** [comment]
Thanks @LecrisUT!
Just to confirm, `pip install git+https://github.com/spglib/spglib --config-settings=cmake.define.SPGLIB_SHARED_LIBS=OFF` works for this, but `pip install spglib --config-settings=cmake.define.SPGLIB_SHARED_LIBS=OFF` doesn't

**Pull request merged**